db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -848,11 +848,17 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
   if (options.change_level) {
     ROCKS_LOG_INFO(immutable_db_options_.info_log,
                    "[RefitLevel] waiting for background threads to stop");
+    DisableManualCompaction();
```

Comment:
I can see this would prevent some issues, but is only disabling manual compaction in refitting phase enough? If it is not, should we disable it in the whole CompactRange() phase if change_level is true?

Reply:
This PR ensures any concurrent manual compaction commits its files before we do `ReFitLevel()`. We know `ReFitLevel()` has a bug in identifying which levels are empty -- but that bug appears independent of whether those levels became nonempty due to a manual compaction or automatic compaction. I don't know of any bug at this point that is specific to manual compaction running concurrently with a `change_level` `CompactRange()` and committing prior to `ReFitLevel()`.

That said, I wouldn't mind if we disabled all flushes, compactions, and ingestions for the entire duration of a `change_level` `CompactRange()`. Then we can simplify things a lot. But it's a significant change in terms of code and behavior.

Reply:
It makes sense. Disabling manual compaction here and check level emptiness might be good enough.

Reply:
Side note: `DBImpl::refitting_level_` (see https://github.com/facebook/rocksdb/blob/master/db/db_impl/db_impl_compaction_flush.cc#L1300-L1306) can probably be removed now?

Reply:
Could it be possible two manual compactions both finish all their `RunManualCompaction()` steps before either one calls `DisableManualCompaction()`? In that case the calls to `DisableManualCompaction()` do not appear effective since manual compactions are only registered during `RunManualCompaction()` steps. Then, the two `ReFitLevel()`s can execute in parallel. I do hope we can consolidate these eventually, just am not sure we're there yet.